src/slide-image-source.cc

```diff
@@ -0,0 +1,164 @@
+
+#include "framebuffer.h"
+#include "terminal-canvas.h"
+
+#include <cassert>
+#include <cstdio>
+#include <unistd.h>
+
+#include <limits>
+#include <string>
+#include <utility>
+
+#include <openslide.h>
+
+extern "C" {
+#  include <libswscale/swscale.h>
+}
+
+namespace timg {
+const char *SlideImageSource::VersionInfo() {
+    return openslide_get_version();
+}
+
+static void dummy_log(void *, int, const char *, va_list) {}
+
+// Scoped clean-up of c objects.
+struct ScopeGuard {
+    ScopeGuard(std::function<void()> f) : f_(f) {}
+    ~ScopeGuard() { f_(); }
+    std::function<void()> f_;
+};
+
+static inline bool invalid_dimensions(int64_t width, int64_t height) {
+    return (width < 0 || height < 0) ||
+        (width > std::numeric_limits<int>::max() ||
+         height > std::numeric_limits<int>::max());
+}
+
+bool SlideImageSource::LoadAndScale(const DisplayOptions &opts, int, int) {
+    options_ = opts;
+    if (opts.scroll_animation ||
+        filename() == "/dev/stdin" || filename() == "-") {
+        return false;  // Not dealing with these now.
+    }
+
+    // Check if this format is supported
+    openslide_t *osr = openslide_open(filename().c_str());
+    if (!osr)
+        return false;
+
+    // Automatically cleanup openslide context
+    ScopeGuard s([osr]() { openslide_close(osr); });
+
+    std::unique_ptr<timg::Framebuffer> source_image;
+    int64_t width, height, l0_width, l0_height;
+    int target_width;
+    int target_height;
+
+    // We need to use one of the layers of the pyramid.
+    // Get native the highest resolution image.
+    openslide_get_level0_dimensions(osr, &l0_width, &l0_height);
+    if (invalid_dimensions(l0_width, l0_height))
+        return false;
+
+    // Get the target size
+    CalcScaleToFitDisplay(l0_width, l0_height, opts, false,
+                          &target_width, &target_height);
+
+    // First check if there's a thumbnail with enough resolution.
+    const char* const* associated_image_names = openslide_get_associated_image_names(osr);
+    while (*associated_image_names) {
+        const char *name = *associated_image_names;
+        openslide_get_associated_image_dimensions(osr, name, &width, &height);
+        if (invalid_dimensions(width, height))
+            return false;
+
+        if (strcmp(name, "thumbnail") == 0) {
+            // If the thumbnail width is smaller than the target,
+            // means we can sample from a larger image.
+            if (width < target_width)
+                break;
+
+            source_image.reset(new timg::Framebuffer(width, height));
+            openslide_read_associated_image(
+                osr, name, (uint32_t *) source_image->begin());
+            break;
+        }
+        associated_image_names++;
+    }
+
+    if (!source_image) {
+        // Get the best layer to downsample with scaling
+        // computed from layer 0 (the highest resolution one).
+        const double scaling = (double) l0_width / target_width;
```

Comment:
is it intentional that this is the inverse ? If `level0_width` is 10000 and our target width is 100, then this calculates a scaling of 100, but I'd expect 1/100 is the scale we're shooting for ?
If it is intentional, we probably need a comment.

Reply:
(solved), openslide wants as "scaling" a downscaling factor.